### ISSUE TITLE
Reactivity gets broken

### ISSUE THREAD

**Issue by XantreGodlike** [software-mansion/react-native-reanimated#4627]
### Description

Reactivity gets broken after many sharedValue assignments.

How its gets broken in reality
```tsx
const Component () => {
  // resume trip mutation
  const resume = () =>
    resumeTrip({ trip, userLevel }).then(
      ({ location$, speed$, region$, stats$, accuracy$ }) => {
        speed$.subscribe((speedValue) => {
          logObservable(speedValue);
          speed.value = speedValue;
          console.log('current speed value', speed.value);
        });
        return location$;
     }

  useEffect(() => {
    const resumeResult = resume();

    return () => {
      resumeResult.then(
        (location$) => !location$.closed && location$.complete(),
      );
    };
  }, [])

  const smoothSpeed = useDerivedValue(() =>
    withSpring(speed.value, {
      overshootClamping: true,
    }),
  );
  useAnimatedReaction(
    () => speed.value,
    (v) => runOnJS(logShared)(v),
  );
  useAnimatedReaction(
    () => smoothSpeed.value,
    (v) => runOnJS(logSmoothed)(v),
  );
  // ....
}
```

```
//... really a lot updates
4:02:12 PM | console | DEBUG : location$ 
{
  "isError": false,
  "data": {
    "latitude": 37.3339359,
    "longitude": -121.8462981,
    "speed": 329.6436218261719,
    "timestamp": 1687784532633,
    "accuracy": 5
  }
}
4:02:12 PM | console | DEBUG : speed from 'shared value' 
329.8852066040039
4:02:12 PM | console | DEBUG : speed from 'smoothed shared value' 
358.7257736206055
4:02:12 PM | console | DEBUG : speed from 'smoothed shared value' 
358.7203564358598
4:02:12 PM | console | DEBUG : speed from 'smoothed shared value' 
358.25722400677176
4:02:12 PM | console | DEBUG : speed from 'smoothed shared value' 
357.1409318051735
4:02:12 PM | console | DEBUG : speed from 'smoothed shared value' 
355.50024763890457
4:02:12 PM | console | DEBUG : speed from 'smoothed shared value' 
353.4575573793984
4:02:12 PM | console | DEBUG : speed from 'smoothed shared value' 
351.1267205343232
4:02:12 PM | console | DEBUG : speed from 'smoothed shared value' 
348.611472680529
4:02:12 PM | console | DEBUG : speed from 'smoothed shared value' 
346.0043315921523
4:02:12 PM | console | DEBUG : speed from 'smoothed shared value' 
343.38595765388936
4:02:12 PM | console | DEBUG : speed from 'smoothed shared value' 
340.8249151512753
4:02:12 PM | console | DEBUG : speed from 'smoothed shared value' 
338.3777789873506
4:02:12 PM | console | DEBUG : speed from 'smoothed shared value' 
336.08953106943056
4:02:12 PM | console | DEBUG : speed from 'smoothed shared value' 
333.99419178415747
4:02:13 PM | console | DEBUG : speed from 'smoothed shared value' 
332.1156343430398
4:02:13 PM | console | DEBUG : speed from 'smoothed shared value' 
330.46853314391694
4:02:13 PM | console | DEBUG : speed from 'smoothed shared value' 
329.8852066040039
4:02:13 PM | console | DEBUG : smoothSpeed$ 
349.1003952026367
4:02:13 PM | console | DEBUG : speed from 'observable subscriber' 
352.1003952026367
4:02:13 PM | console | DEBUG : current speed value 
329.8852066040039
4:02:13 PM | console | DEBUG : location$ 
{
  "isError": false,
  "data": {
    "latitude": 37.3346989,
    "longitude": -121.8470679,
    "speed": 390.5029632568359,
    "timestamp": 1687784533612,
    "accuracy": 5
  }
}
4:02:13 PM | console | DEBUG : speed from 'shared value' 
352.1003952026367
4:02:13 PM | console | DEBUG : speed from 'smoothed shared value' 
329.8852066040039
4:02:13 PM | console | DEBUG : speed from 'smoothed shared value' 
329.92806680510944
4:02:13 PM | console | DEBUG : speed from 'smoothed shared value' 
330.42535494464823
4:02:13 PM | console | DEBUG : speed from 'smoothed shared value' 
331.3996040438667
4:02:13 PM | console | DEBUG : speed from 'smoothed shared value' 
334.3920041640773
4:02:13 PM | console | DEBUG : speed from 'smoothed shared value' 
336.2316727686217
4:02:13 PM | console | DEBUG : speed from 'smoothed shared value' 
338.1940031864503
4:02:13 PM | console | DEBUG : speed from 'smoothed shared value' 
340.21007172522104
4:02:13 PM | console | DEBUG : speed from 'smoothed shared value' 
342.22016395315836
4:02:13 PM | console | DEBUG : speed from 'smoothed shared value' 
344.17388525274987
4:02:13 PM | console | DEBUG : speed from 'smoothed shared value' 
346.03001649550714
4:02:13 PM | console | DEBUG : speed from 'smoothed shared value' 
347.7561576454326
4:02:13 PM | console | DEBUG : speed from 'smoothed shared value' 
349.3282009519298
4:02:13 PM | console | DEBUG : speed from 'smoothed shared value' 
350.72967333523445
4:02:13 PM | console | DEBUG : speed from 'smoothed shared value' 
351.9509848382334
4:02:13 PM | console | DEBUG : speed from 'smoothed shared value' 
352.1003952026367
4:02:14 PM | console | DEBUG : smoothSpeed$ 
386.873698425293
4:02:14 PM | console | DEBUG : speed from 'observable subscriber' 
389.873698425293
4:02:14 PM | console | DEBUG : current speed value 
352.1003952026367
4:02:14 PM | console | DEBUG : location$ 
{
  "isError": false,
  "data": {
    "latitude": 37.3354413,
    "longitude": -121.8478073,
    "speed": 403.5308807373047,
    "timestamp": 1687784534561,
    "accuracy": 5
  }
}
4:02:14 PM | console | DEBUG : speed from 'shared value' 
389.873698425293
4:02:14 PM | console | DEBUG : speed from 'smoothed shared value' 
352.1003952026367
4:02:14 PM | console | DEBUG : speed from 'smoothed shared value' 
353.499574971087
4:02:14 PM | console | DEBUG : smoothSpeed$ 
396.13024749755857
4:02:14 PM | console | DEBUG : speed from 'observable subscriber' 
399.13024749755857
4:02:14 PM | console | DEBUG : current speed value 
389.873698425293
4:02:15 PM | console | DEBUG : speed from 'shared value' 
399.13024749755857
4:02:15 PM | console | DEBUG : speed from 'smoothed shared value' 
364.9790443403077
4:02:15 PM | console | DEBUG : smoothSpeed$ 
399.17433471679686
4:02:15 PM | console | DEBUG : speed from 'observable subscriber' 
402.17433471679686
4:02:15 PM | console | DEBUG : current speed value 
399.13024749755857
4:02:15 PM | console | DEBUG : location$ 
{
  "isError": false,
  "data": {
    "latitude": 37.3361287,
    "longitude": -121.8484764,
    "speed": 401.7902069091797,
    "timestamp": 1687784535472,
    "accuracy": 5
  }
}
4:02:16 PM | console | DEBUG : smoothSpeed$ 
411.9033737182617
4:02:16 PM | console | DEBUG : speed from 'observable subscriber' 
414.9033737182617
4:02:16 PM | console | DEBUG : current speed value 
402.17433471679686
4:02:16 PM | console | DEBUG : location$ 
{
  "isError": false,
  "data": {
    "latitude": 37.3369067,
    "longitude": -121.849253,
    "speed": 424.9225250244141,
    "timestamp": 1687784536396,
    "accuracy": 5
  }
}
4:02:17 PM | console | DEBUG : smoothSpeed$ 
352.7412551879883
4:02:17 PM | console | DEBUG : speed from 'observable subscriber' 
355.7412551879883
4:02:17 PM | console | DEBUG : current speed value 
414.9033737182617
4:02:17 PM | console | DEBUG : location$ 
{
  "isError": false,
  "data": {
    "latitude": 37.3375992,
    "longitude": -121.8499199,
    "speed": 288.27075805664066,
    "timestamp": 1687784537427,
    "accuracy": 5
  }
}
4:02:17 PM | console | DEBUG : smoothSpeed$ 
332.85628967285163
4:02:17 PM | console | DEBUG : speed from 'observable subscriber' 
335.85628967285163
4:02:17 PM | console | DEBUG : current speed value 
355.7412551879883
4:02:18 PM | console | DEBUG : smoothSpeed$ 
337.4267761230469
4:02:18 PM | console | DEBUG : speed from 'observable subscriber' 
340.4267761230469
4:02:18 PM | console | DEBUG : current speed value 
335.85628967285163
4:02:18 PM | console | DEBUG : location$ 
{
  "isError": false,
  "data": {
    "latitude": 37.3382888,
    "longitude": -121.850587,
    "speed": 357.5024780273438,
    "timestamp": 1687784538349,
    "accuracy": 5
  }
}
4:02:19 PM | console | DEBUG : smoothSpeed$ 
376.7831451416016
4:02:19 PM | console | DEBUG : speed from 'observable subscriber' 
379.7831451416016
4:02:19 PM | console | DEBUG : current speed value 
340.4267761230469
4:02:19 PM | console | DEBUG : location$ 
{
  "isError": false,
  "data": {
    "latitude": 37.3390586,
    "longitude": -121.8513275,
    "speed": 404.60089416503905,
    "timestamp": 1687784539270,
    "accuracy": 5
  }
}
4:02:20 PM | console | DEBUG : smoothSpeed$ 
403.66138000488286
4:02:20 PM | console | DEBUG : speed from 'observable subscriber' 
406.66138000488286
4:02:20 PM | console | DEBUG : current speed value 
379.7831451416016
4:02:20 PM | console | DEBUG : location$ 
{
  "isError": false,
  "data": {
    "latitude": 37.3398045,
    "longitude": -121.8520131,
    "speed": 418.421337890625,
    "timestamp": 1687784540173,
    "accuracy": 5
  }
}
4:02:20 PM | console | DEBUG : smoothSpeed$ 
402.02071380615234
4:02:20 PM | console | DEBUG : speed from 'observable subscriber' 
405.02071380615234
4:02:20 PM | console | DEBUG : current speed value 
406.66138000488286
4:02:21 PM | console | DEBUG : smoothSpeed$ 
403.4450500488282
4:02:21 PM | console | DEBUG : speed from 'observable subscriber' 
406.4450500488282
4:02:21 PM | console | DEBUG : current speed value 
405.02071380615234
4:02:21 PM | console | DEBUG : location$ 
{
  "isError": false,
  "data": {
    "latitude": 37.3404601,
    "longitude": -121.8526374,
    "speed": 407.2650512695313,
    "timestamp": 1687784541079,
    "accuracy": 5
  }
}
4:02:22 PM | console | DEBUG : smoothSpeed$ 
349.73502044677736
4:02:22 PM | console | DEBUG : speed from 'observable subscriber' 
352.73502044677736
4:02:22 PM | console | DEBUG : current speed value 
406.4450500488282
4:02:22 PM | console | DEBUG : location$ 
{
  "isError": false,
  "data": {
    "latitude": 37.3414734,
    "longitude": -121.853627,
    "speed": 297.8843719482422,
    "timestamp": 1687784542093,
    "accuracy": 5
  }
}
4:02:23 PM | console | DEBUG : smoothSpeed$ 
355.780307006836
4:02:23 PM | console | DEBUG : speed from 'observable subscriber' 
358.780307006836
4:02:23 PM | console | DEBUG : current speed value 
352.73502044677736
4:02:23 PM | console | DEBUG : smoothSpeed$ 
361.27954559326173
4:02:23 PM | console | DEBUG : speed from 'observable subscriber' 
364.27954559326173
4:02:23 PM | console | DEBUG : current speed value 
352.73502044677736
4:02:23 PM | console | DEBUG : location$ 
{
  "isError": false,
  "data": {
    "latitude": 37.3421352,
    "longitude": -121.8542665,
    "speed": 371.7869567871094,
    "timestamp": 1687784543048,
    "accuracy": 5
  }
}
4:02:23 PM | console | DEBUG : Request url=/trips/track-location traceId=6b8accef-d8f5-4135-8852-0479c637efb8
4:02:23 PM | console | DEBUG : Request url=/trips/track-location traceId=591b55f3-abf6-4f25-aaa8-8d6be0fe1748
4:02:24 PM | console | DEBUG : smoothSpeed$ 
400.9810043334961
4:02:24 PM | console | DEBUG : speed from 'observable subscriber' 
403.9810043334961
4:02:24 PM | console | DEBUG : current speed value 
364.27954559326173
4:02:24 PM | console | DEBUG : location$ 
{
  "isError": false,
  "data": {
    "latitude": 37.3429966,
    "longitude": -121.8550986,
    "speed": 428.36536560058596,
    "timestamp": 1687784543991,
    "accuracy": 5
  }
}
4:02:24 PM | console | DEBUG : smoothSpeed$ 
427.07728271484376
4:02:24 PM | console | DEBUG : speed from 'observable subscriber' 
430.07728271484376
4:02:24 PM | console | DEBUG : current speed value 
403.9810043334961
4:02:24 PM | console | DEBUG : location$ 
{
  "isError": false,
  "data": {
    "latitude": 37.3438109,
    "longitude": -121.8558916,
    "speed": 444.64866943359374,
    "timestamp": 1687784544912,
    "accuracy": 5
  }
}
4:02:25 PM | console | DEBUG : smoothSpeed$ 
439.6367752075196
4:02:25 PM | console | DEBUG : speed from 'observable subscriber' 
442.6367752075196
4:02:25 PM | console | DEBUG : current speed value 
430.07728271484376
4:02:25 PM | console | DEBUG : smoothSpeed$ 
440.81865234375005
4:02:25 PM | console | DEBUG : speed from 'observable subscriber' 
443.81865234375005
4:02:25 PM | console | DEBUG : current speed value 
442.6367752075196
4:02:25 PM | console | DEBUG : location$ 
{
  "isError": false,
  "data": {
    "latitude": 37.3446448,
    "longitude": -121.856692,
    "speed": 440.0526489257813,
    "timestamp": 1687784545893,
    "accuracy": 5
  }
}
4:02:26 PM | console | DEBUG : smoothSpeed$ 
375.26439056396487
4:02:26 PM | console | DEBUG : speed from 'observable subscriber' 
378.26439056396487
4:02:26 PM | console | DEBUG : current speed value 
443.81865234375005
4:02:27 PM | console | DEBUG : location$ 
{
  "isError": false,
  "data": {
    "latitude": 37.3456791,
    "longitude": -121.857689,
    "speed": 310.47613220214845,
    "timestamp": 1687784546896,
    "accuracy": 5
  }
}
4:02:27 PM | console | DEBUG : smoothSpeed$ 
360.3027282714844
4:02:27 PM | console | DEBUG : speed from 'observable subscriber' 
363.3027282714844
4:02:27 PM | console | DEBUG : current speed value 
378.26439056396487
4:02:27 PM | console | DEBUG : location$ 
{
  "isError": false,
  "data": {
    "latitude": 37.3462758,
    "longitude": -121.8582699,
    "speed": 366.93715209960936,
    "timestamp": 1687784547809,
    "accuracy": 5
  }
}
4:02:28 PM | console | DEBUG : smoothSpeed$ 
459.1489517211914
4:02:28 PM | console | DEBUG : speed from 'observable subscriber' 
462.1489517211914
4:02:28 PM | console | DEBUG : current speed value 
363.3027282714844
4:02:28 PM | console | DEBUG : smoothSpeed$ 
529.2429748535156
4:02:28 PM | console | DEBUG : speed from 'observable subscriber' 
532.2429748535156
4:02:28 PM | console | DEBUG : current speed value 
462.1489517211914
4:02:28 PM | console | DEBUG : location$ 
{
  "isError": false,
  "data": {
    "latitude": 37.3471785,
    "longitude": -121.8592298,
    "speed": 556.0528381347657,
    "timestamp": 1687784548801,
    "accuracy": 5
  }
}
4:02:29 PM | console | DEBUG : smoothSpeed$ 
519.9185943603516
4:02:29 PM | console | DEBUG : speed from 'observable subscriber' 
522.9185943603516
4:02:29 PM | console | DEBUG : current speed value 
532.2429748535156
4:02:29 PM | console | DEBUG : location$ 
{
  "isError": false,
  "data": {
    "latitude": 37.3480143,
    "longitude": -121.8598862,
    "speed": 479.07493286132814,
    "timestamp": 1687784549736,
    "accuracy": 5
  }
}
4:02:30 PM | console | DEBUG : smoothSpeed$ 
476.5770492553711
4:02:30 PM | console | DEBUG : speed from 'observable subscriber' 
479.5770492553711
4:02:30 PM | console | DEBUG : current speed value 
522.9185943603516
4:02:30 PM | console | DEBUG : location$ 
{
  "isError": false,
  "data": {
    "latitude": 37.3488299,
    "longitude": -121.8605858,
    "speed": 448.41986389160155,
    "timestamp": 1687784550687,
    "accuracy": 5
  }
}
4:02:31 PM | console | DEBUG : smoothSpeed$ 
452.23979644775386
4:02:31 PM | console | DEBUG : speed from 'observable subscriber' 
455.23979644775386
4:02:31 PM | console | DEBUG : current speed value 
479.5770492553711
4:02:31 PM | console | DEBUG : smoothSpeed$ 
442.5510543823243
4:02:31 PM | console | DEBUG : speed from 'observable subscriber' 
445.5510543823243
4:02:31 PM | console | DEBUG : current speed value 
455.23979644775386
4:02:31 PM | console | DEBUG : location$ 
{
  "isError": false,
  "data": {
    "latitude": 37.3496455,
    "longitude": -121.8612993,
    "speed": 438.4012390136719,
    "timestamp": 1687784551624,
    "accuracy": 5
  }
}
4:02:32 PM | console | DEBUG : smoothSpeed$ 
382.9540328979492
4:02:32 PM | console | DEBUG : speed from 'observable subscriber' 
385.9540328979492
4:02:32 PM | console | DEBUG : current speed value 
445.5510543823243
4:02:32 PM | console | DEBUG : location$ 
{
  "isError": false,
  "data": {
    "latitude": 37.3504442,
    "longitude": -121.8620031,
    "speed": 325.0267822265625,
    "timestamp": 1687784552595,
    "accuracy": 5
  }
}
4:02:33 PM | console | DEBUG : smoothSpeed$ 
478.53196105957034
4:02:33 PM | console | DEBUG : speed from 'observable subscriber' 
481.53196105957034
4:02:33 PM | console | DEBUG : current speed value 
385.9540328979492
4:02:33 PM | console | DEBUG : location$ 
{
  "isError": false,
  "data": {
    "latitude": 37.3512058,
    "longitude": -121.8628698,
    "speed": 594.245654296875,
    "timestamp": 1687784553519,
    "accuracy": 5
  }
}
4:02:34 PM | console | DEBUG : smoothSpeed$ 
526.8780944824218
4:02:34 PM | console | DEBUG : speed from 'observable subscriber' 
529.8780944824218
4:02:34 PM | console | DEBUG : current speed value 
481.53196105957034
4:02:34 PM | console | DEBUG : smoothSpeed$ 
537.9374084472656
4:02:34 PM | console | DEBUG : speed from 'observable subscriber' 
540.9374084472656
4:02:34 PM | console | DEBUG : current speed value 
529.8780944824218
4:02:34 PM | console | DEBUG : location$ 
{
  "isError": false,
  "data": {
    "latitude": 37.3519637,
    "longitude": -121.8635518,
    "speed": 511.62615966796875,
    "timestamp": 1687784554426,
    "accuracy": 5
  }
}
```

### Steps to reproduce

1. Try to update sharedValue many times
2. Use it with some reactive hooks like `useDerivedValue` or `useAnimatedReaction`
3. Wait before reactivity stops working (when it happens can vary, for me it happened on 700/1313/2922/7693 iteration)

```
 LOG  sum 400245.0647423127
 LOG  update 674
 LOG  sum 8843.48881068066
 LOG  update 675
 LOG  sum 470395.17651071766
 LOG  update 676
 LOG  sum 489927.91066198534
 LOG  update 677
 LOG  sum 159372.4616868039
 LOG  update 678
 LOG  sum 498823.06357880577
 LOG  update 679
 LOG  sum 356568.6609531895
 LOG  update 680
 LOG  sum 86998.1732637743
 LOG  update 681
 LOG  sum 245003.9161713042
 LOG  update 682
 LOG  sum 407489.5091257119
 LOG  update 683
 LOG  sum 196853.49674187286
 LOG  update 684
 LOG  sum 265455.51339213515
 LOG  update 685
 LOG  sum 25022.637878140584
 LOG  update 686
 LOG  sum 281274.3729307794
 LOG  update 687
 LOG  sum 264803.5465995123
 LOG  update 688
 LOG  sum 336257.79310684116
 LOG  update 689
 LOG  sum 44757.63336496609
 LOG  update 690
 LOG  sum 372227.5567404828
 LOG  update 691
 LOG  sum 259575.91936250404
 LOG  update 692
 LOG  sum 255207.67538324717
 LOG  update 693
 LOG  sum 282571.46066978446
 LOG  update 694
 LOG  sum 417680.20640934806
 LOG  update 695
 LOG  sum 7604.672532269968
 LOG  update 696
 LOG  sum 209347.7232981703
 LOG  update 697
 LOG  sum 111391.8726154284
 LOG  update 698
 LOG  update 699
 LOG  sum 275758.93085384794
 LOG  sum 398365.3486152557
 LOG  update 700
 LOG  sum 418636.2073295794
 LOG  update 701
 LOG  update 702
 LOG  update 703
 LOG  update 704
 LOG  update 705
 LOG  update 706
 LOG  update 707
 LOG  update 708
 LOG  update 709
 LOG  update 710
 LOG  update 711
 LOG  update 712
 LOG  update 713
 LOG  update 714
 LOG  update 715
 LOG  update 716
 LOG  update 717
 LOG  update 718
 LOG  update 719
 LOG  update 720
 LOG  update 721
 LOG  update 722
```

### Snack or a link to a repository

https://github.com/XantreGodlike/reanimated-broken-reactivity

### Reanimated version

3.3.0

### React Native version

0.71.7

### Platforms

Android

### JavaScript runtime

Hermes

### Workflow

Expo bare workflow

### Architecture

Paper (Old Architecture)

### Build type

None

### Device

Android emulator

### Acknowledgements

Yes

**Comment by tomekzaw** [comment]
Hey @XantreGodlike, thanks for reporting this issue. Can you please check if PR #4633 fixes it?

**Comment by XantreGodlike** [comment]
I will try tomorrow

**Comment by XantreGodlike** [comment]
I've created new branch with applied patch, and this bug still reproducing
https://github.com/XantreGodlike/reanimated-broken-reactivity/tree/check-use-ref-hypotesis

**Comment by XantreGodlike** [comment]
I think its unrelated with useAnimatedReaction, it about reactivity system, useDerivedValue stopping to execute too. Values changes, but dependecies is not updating

**Comment by XantreGodlike** [comment]
@tomekzaw what do you think about it?

**Comment by tomekzaw** [comment]
@XantreGodlike I haven't had time to investigate this issue yet.

**Comment by XantreGodlike** [comment]
Reactivity in 2.17.0 seems to be work fine

**Comment by XantreGodlike** [comment]
2.17.0: ok https://github.com/XantreGodlike/reanimated-broken-reactivity/tree/reanimated2-check
3.1.0: not ok https://github.com/XantreGodlike/reanimated-broken-reactivity/tree/reanimated3.1

**Comment by XantreGodlike** [comment]
It's critical bug, and blocks our bug fix, because we are using `react-native-skia` that uses `react-native-reanimated` v3

**Comment by XantreGodlike** [comment]
@tomekzaw do you want I try to implement actual runtime reactivity system instead, this strange one. I think preact signals would look great for this purpose

**Comment by Eonfuzz** [comment]
Any fixes found for this issue? I'm experiencing it on a simple `runOnJS(someMemoisedFunction)()`

**Comment by XantreGodlike** [comment]
Which version of reanimated do you use? Seems to be it was fixed in 3.4

**Comment by Latropos** [comment]
Closing since the problem is allegedly fixed in 3.4
